mlrun/utils/http.py

```diff
@@ -0,0 +1,114 @@
+from . import logger
+
+DEFAULT_RETRY_COUNT = 3
+DEFAULT_RETRY_BACKOFF = 1
+
+
+class HTTPSessionWithRetry(requests.Session):
+
+    # make sure to only add exceptions that are raised early in the request. For example, ConnectionError can be raised
+    # during the handling of a request, and therefore should not be retried, as the request might not be idempotent.
+    HTTP_RETRYABLE_EXCEPTIONS = {
+        # ConnectionResetError is raised when the server closes the connection prematurely during TCP handshake.
+        ConnectionResetError: ["Connection reset by peer", "Connection aborted"],
+        # "Connection aborted" and "Connection refused" happen when the server doesn't respond at all.
+        ConnectionError: ["Connection aborted", "Connection refused"],
+        ConnectionRefusedError: ["Connection refused"],
+    }
+
+    def __init__(
+        self,
+        max_retries=DEFAULT_RETRY_COUNT,
+        retry_backoff_factor=DEFAULT_RETRY_BACKOFF,
+        retry_on_exception=True,
+        retry_on_status=True,
+        retry_on_post=False,
+    ):
+        super().__init__()
+
+        self.max_retries = max_retries
+        self.retry_backoff_factor = retry_backoff_factor
+        self.retry_on_exception = retry_on_exception
+
+        if retry_on_status:
+            http_adapter = requests.adapters.HTTPAdapter(
+                max_retries=urllib3.util.retry.Retry(
+                    total=self.max_retries,
+                    backoff_factor=self.retry_backoff_factor,
+                    status_forcelist=[500, 502, 503, 504],
+                    method_whitelist=False
+                    if retry_on_post
+                    else urllib3.util.retry.Retry.DEFAULT_METHOD_WHITELIST,
+                    # we want to retry but not to raise since we do want that last response (to parse details on the
+                    # error from response body) we'll handle raising ourselves
+                    raise_on_status=False,
+                ),
+                pool_maxsize=int(config.httpdb.max_workers),
+            )
+
+            self.mount("http://", http_adapter)
+            self.mount("https://", http_adapter)
+
+    def request(self, method, url, **kwargs):
+        retry_count = 0
+        while True:
+            try:
+                response = super().request(method, url, **kwargs)
+                return response
+            except tuple(self.HTTP_RETRYABLE_EXCEPTIONS.keys()) as exc:
+                if not self.retry_on_exception:
+                    logger.warning(
+                        f"{method} {url} request failed, http retries disabled, raising exception: {exc}",
+                        exception_type=type(exc),
+                        exception_message=str(exc),
+                        retry_interval=self.retry_backoff_factor,
+                        retry_count=retry_count,
+                        max_retries=self.max_retries,
+                    )
+                    raise exc
```

Comment:
This log block is repeated 2.5 times - twice here and 3rd time with less parameters if it's non-retryable (but could have the same set of parameters there as well). Should make this an internal function or just extract it to a function, rather than repeating the same code again and again.